src/python/pants/backend/python/util_rules/faas.py

```diff
@@ -430,16 +430,29 @@ async def infer_runtime_platforms(request: RuntimePlatformsRequest) -> RuntimePl
         file_name = next(
             rt.file_name()
             for rt in request.runtime.known_runtimes
-            if version == (rt.major, rt.minor)
-            and (request.architecture is None or request.architecture.value == rt.architecture)
+            if version == (rt.major, rt.minor) and request.architecture.value == rt.architecture
         )
     except StopIteration:
-        # Not a known runtime, so fallback to just passing a platform
-        # TODO: maybe this should be an error, and we require users to specify
-        # complete_platforms themselves?
-        return RuntimePlatforms(
-            interpreter_version=version,
-            pex_platforms=PexPlatforms((_format_platform_from_major_minor(*version),)),
+        # Not a known runtime, so error out.
+        raise ValueError(
+            softwrap(
+                f"""\
+                Could not find a known runtime for the given Python version and machine architecture!
+
+                Python version: {version}
+                Machine architecture: {request.architecture.value}
+
+                Known runtimes:
+                {request.runtime.known_runtimes}
+
+                To fix, please generate a `complete_platforms` file for the given Python version and
+                machine architecture, or specify a runtime that is known to Pants.
+
+                You can follow the instructions at https://github.com/pantsbuild/pants/discussions/18756
+                to generate a `complete_platforms` file for your Python version and machine
+                architecture.
```

Comment:
This still needs a little bit of work. I'll try to massage the output into something a little more user-friendly / readable.

Reply:
I'm big fan of this change! Thank you for starting the removal.

However, I'm a bit nervous about this being a breaking change for some reason we haven't thought of (i.e. despite having known runtimes covering what we believe are the "live" environments).

Thus, can I suggest:

1. doing this in a separate pull request with its own release notes
2. doing it via a deprecation (a call to `pants.base.deprecated.warn_or_error` before the `return`, without removing the `return`, maybe)

That'd let us get the GCF improvement in independently of work-shopping this error.

What do you think?

Reply:
Sure, that seems quite reasonable. I did like bundling this change in to the current PR because it allowed me to find and fix breaking tests but we can definitely split it out for user facing changes.

Reply:
> Sure, that seems quite reasonable.

Thanks!

> I did like bundling this change in to the current PR because it allowed me to find and fix breaking tests

Ah, yeah, I see how that's helpful. Nice one! 👍